### DISCUSSION TITLE
User Roles within Auth-Token

### DISCUSSION THREAD

**Discussion by muelbr** [zitadel/zitadel#1979] [Q&A] [answered]
Hi there

I have a project with some user roles (assigned to some users) and I enabled  "Assert Roles on Authentication" (project) and "user roles inside ID token" (application). But I am not able to get any roles from the token - do I get it wrong?

I would like to get the roles from the auth token on the backend for Spring security allowing REST calls some role based access. I would like to avoid some extra calls to Zitadel on a REST call get the user claims...

Regards
BM

**Comment by livio-a** [comment, marked as answer]
Hi @muelbr 

From what I read, you're doing everything right. I just tested and could reproduce your problem.

There are two problems at the moment:
The UI descriptions need improvement or are even misleading. I'll improve them and add a section in the docs as well.

In short there are multiple options to get (certain) roles of an authenticated user:
Request certain roles by scopes or have all roles returned by checking the "Assert Roles on Authentication" (project)
Either way roles will be returned on userinfo / introspection.
If you want them to be returned on id_token or access_token you will have to  check "user roles inside ID token" (application) or "user roles inside access token". The access token option will only be displayed if the token type is set to JWT.

Second problem is that at the moment it's never displayed.

I'll create an issue and fix this today. In the meantime you could set the access_token_role_assertion via API.

**Reply by livio-a** [reply]
hey @muelbr 

We just deployed a fix. "User roles to access token" is not displayed again and we tried to improve the descriptions.
Docs will be updated soon as well.

Let me know if anything is unclear or does not work

Regards
Livio

**Reply by muelbr** [reply]
Thanks a lot - it is currently working... 

Regards
BM
